// angular-maps/dist/src/components/map-polyline.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var polyline_service_1 = require("./../services/polyline.service");
var infobox_1 = require("./infobox");
var polylineId = 0;
/**
 *
 * MapPolylineDirective renders a polyline inside a {@link MapComponent}.
 *
 * ### Example
 * ```typescript
 * import {Component} from '@angular/core';
 * import {MapComponent, MapPolylineDirective} from '...';
 *
 * @Component({
 *  selector: 'my-map,
 *  styles: [`
 *   .map-container { height: 300px; }
 * `],
 * template: `
 *   <x-map [Latitude]="lat" [Longitude]="lng" [Zoom]="zoom">
 *      <x-map-polyline [Paths]="path"></x-map-polyline>
 *   </x-map>
 * `
 * })
 * ```
 *
 *
 * @export
 * @class MapPolylineDirective
 * @implements {OnDestroy}
 * @implements {OnChanges}
 * @implements {AfterContentInit}
 */
var MapPolylineDirective = (function () {
    ///
    /// Constructor
    ///
    /**
     * Creates an instance of MapPolylineDirective.
     * @param {PolylineManager} _polylineManager
     *
     * @memberof MapPolylineDirective
     */
    function MapPolylineDirective(_polylineService, _containerRef) {
        this._polylineService = _polylineService;
        this._containerRef = _containerRef;
        ///
        /// Field declarations
        ///
        this._inCustomLayer = false;
        this._addedToService = false;
        this._events = [];
        /**
         * Gets or sets whether this Polyline handles mouse events.
         *
         * @type {boolean}
         * @memberof MapPolylineDirective
         */
        this.Clickable = true;
        /**
         * If set to true, the user can drag this shape over the map.
         *
         * @type {boolean}
         * @memberof MapPolylineDirective
         */
        this.Draggable = false;
        /**
         * If set to true, the user can edit this shape by dragging the control
         * points shown at the vertices and on each segment.
         *
         * @type {boolean}
         * @memberof MapPolylineDirective
         */
        this.Editable = false;
        /**
         * When true, edges of the polyline are interpreted as geodesic and will
         * follow the curvature of the Earth. When false, edges of the polyline are
         * rendered as straight lines in screen space. Note that the shape of a
         * geodesic polyline may appear to change when dragged, as the dimensions
         * are maintained relative to the surface of the earth. Defaults to false.
         *
         * @type {boolean}
         * @memberof MapPolylineDirective
         */
        this.Geodesic = false;
        /**
         * The ordered sequence of coordinates that designates a polyline.
         * Simple polylines may be defined using a single array of LatLngs. More
         * complex polylines may specify an array of arrays.
         *
         * @type {(Array<ILatLong>}
         * @memberof MapPolylineDirective
         */
        this.Path = [];
        /**
         * Whether to show the title of the polyline as the tooltip on the polygon.
         *
         * @type {boolean}
         * @memberof MapPolylineDirective
         */
        this.ShowTooltip = true;
        ///
        /// Delegate definitions
        ///
        /**
         * This event is fired when the DOM click event is fired on the Polyline.
         *
         *   @type {EventEmitter<MouseEvent>}
         * @memberof MapPolylineDirective
         */
        this.Click = new core_1.EventEmitter();
        /**
         * This event is fired when the DOM dblclick event is fired on the Polyline.
         *
         * @type {EventEmitter<MouseEvent>}
         * @memberof MapPolylineDirective
         */
        this.DblClick = new core_1.EventEmitter();
        /**
         * This event is repeatedly fired while the user drags the polyline.
         *
         * @type {EventEmitter<MouseEvent>}
         * @memberof MapPolylineDirective
         */
        this.Drag = new core_1.EventEmitter();
        /**
         * This event is fired when the user stops dragging the polyline.
         *
         * @type {EventEmitter<MouseEvent>}
         * @memberof MapPolylineDirective
         */
        this.DragEnd = new core_1.EventEmitter();
        /**
         * This event is fired when the user starts dragging the polyline.
         *
         * @type {EventEmitter<MouseEvent>}
         * @memberof MapPolylineDirective
         */
        this.DragStart = new core_1.EventEmitter();
        /**
         * This event is fired when the DOM mousedown event is fired on the Polyline.
         *
         * @type {EventEmitter<MouseEvent>}
         * @memberof MapPolylineDirective
         */
        this.MouseDown = new core_1.EventEmitter();
        /**
         * This event is fired when the DOM mousemove event is fired on the Polyline.
         *
         * @type {EventEmitter<MouseEvent>}
         * @memberof MapPolylineDirective
         */
        this.MouseMove = new core_1.EventEmitter();
        /**
         * This event is fired on Polyline mouseout.
         *
         * @type {EventEmitter<MouseEvent>}
         * @memberof MapPolylineDirective
         */
        this.MouseOut = new core_1.EventEmitter();
        /**
         * This event is fired on Polyline mouseover.
         *
         * @type {EventEmitter<MouseEvent>}
         * @memberof MapPolylineDirective
         */
        this.MouseOver = new core_1.EventEmitter();
        /**
         * This event is fired whe the DOM mouseup event is fired on the Polyline
         *
         * @type {EventEmitter<MouseEvent>}
         * @memberof MapPolylineDirective
         */
        this.MouseUp = new core_1.EventEmitter();
        /**
         * This even is fired when the Polyline is right-clicked on.
         *
         * @type {EventEmitter<MouseEvent>}
         * @memberof MapPolylineDirective
         */
        this.RightClick = new core_1.EventEmitter();
        this._id = polylineId++;
    }
    Object.defineProperty(MapPolylineDirective.prototype, "AddedToService", {
        ///
        /// Property declarations
        ///
        /**
         * Gets whether the polyline has been registered with the service.
         * @readonly
         * @type {boolean}
         * @memberof MapPolylineDirective
         */
        get: function () { return this._addedToService; },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(MapPolylineDirective.prototype, "Id", {
        /**
         * Get the id of the polyline.
         *
         * @readonly
         * @type {number}
         * @memberof MapPolylineDirective
         */
        get: function () { return this._id; },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(MapPolylineDirective.prototype, "IdAsString", {
        /**
         * Gets the id of the polyline as a string.
         *
         * @readonly
         * @type {string}
         * @memberof MapPolylineDirective
         */
        get: function () { return this._id.toString(); },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(MapPolylineDirective.prototype, "InCustomLayer", {
        /**
         * Gets whether the polyline is in a custom layer. See {@link MapLayer}.
         *
         * @readonly
         * @type {boolean}
         * @memberof MapPolylineDirective
         */
        get: function () { return this._inCustomLayer; },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(MapPolylineDirective.prototype, "LayerId", {
        /**
         * gets the id of the Layer the polyline belongs to.
         *
         * @readonly
         * @type {number}
         * @memberof MapPolylineDirective
         */
        get: function () { return this._layerId; },
        enumerable: true,
        configurable: true
    });
    ///
    /// Public methods
    ///
    /**
     * Called after the content intialization of the directive is complete. Part of the ng Component life cycle.
     *
     * @return {void}
     *
     * @memberof MapPolylineDirective
     */
    MapPolylineDirective.prototype.ngAfterContentInit = function () {
        if (this._containerRef.element.nativeElement.parentElement) {
            var parentName = this._containerRef.element.nativeElement.parentElement.tagName;
            if (parentName.toLowerCase() === 'x-map-layer') {
                this._inCustomLayer = true;
                this._layerId = Number(this._containerRef.element.nativeElement.parentElement.attributes['layerId']);
            }
        }
        if (!this._addedToService) {
            this._polylineService.AddPolyline(this);
            this._addedToService = true;
            this.AddEventListeners();
        }
        return;
    };
    /**
     * Called when changes to the databoud properties occur. Part of the ng Component life cycle.
     *
     * @param {{ [propName: string]: SimpleChange }} changes - Changes that have occured.
     * @return {void}
     *
     * @memberof MapPolylineDirective
     */
    MapPolylineDirective.prototype.ngOnChanges = function (changes) {
        if (!this._addedToService) {
            return;
        }
        var o = this.GeneratePolylineChangeSet(changes);
        this._polylineService.SetOptions(this, o);
        if (changes['Path'] && !changes['Path'].isFirstChange()) {
            this._polylineService.UpdatePolyline(this);
        }
    };
    /**
     * Called when the polyline is being destroyed. Part of the ng Component life cycle. Release resources.
     *
     *
     * @memberof MapPolylineDirective
     */
    MapPolylineDirective.prototype.ngOnDestroy = function () {
        this._polylineService.DeletePolyline(this);
        this._events.forEach(function (s) { return s.unsubscribe(); });
        ///
        /// remove event subscriptions
        ///
    };
    ///
    /// Private methods
    ///
    /**
     * Wires up the event receivers.
     *
     * @private
     *
     * @memberof MapPolylineDirective
     */
    MapPolylineDirective.prototype.AddEventListeners = function () {
        var _this = this;
        this._polylineService.CreateEventObservable('click', this).subscribe(function (ev) {
            if (_this._infoBox != null) {
                _this._infoBox.Open(_this._polylineService.GetCoordinatesFromClick(ev));
            }
        });
        var handlers = [
            { name: 'click', handler: function (ev) { return _this.Click.emit(ev); } },
            { name: 'dblclick', handler: function (ev) { return _this.DblClick.emit(ev); } },
            { name: 'drag', handler: function (ev) { return _this.Drag.emit(ev); } },
            { name: 'dragend', handler: function (ev) { return _this.DragEnd.emit(ev); } },
            { name: 'dragstart', handler: function (ev) { return _this.DragStart.emit(ev); } },
            { name: 'mousedown', handler: function (ev) { return _this.MouseDown.emit(ev); } },
            { name: 'mousemove', handler: function (ev) { return _this.MouseMove.emit(ev); } },
            { name: 'mouseout', handler: function (ev) { return _this.MouseOut.emit(ev); } },
            { name: 'mouseover', handler: function (ev) { return _this.MouseOver.emit(ev); } },
            { name: 'mouseup', handler: function (ev) { return _this.MouseUp.emit(ev); } },
            { name: 'rightclick', handler: function (ev) { return _this.RightClick.emit(ev); } },
        ];
        handlers.forEach(function (obj) {
            var os = _this._polylineService.CreateEventObservable(obj.name, _this).subscribe(obj.handler);
            _this._events.push(os);
        });
    };
    /**
     * Generates IPolyline option changeset from directive settings.
     *
     * @private
     * @param {SimpleChanges} changes - {@link SimpleChanges} identifying the changes that occured.
     * @returns {IPolylineOptions} - {@link IPolylineOptions} containing the polyline options.
     *
     * @memberof MapPolylineDirective
     */
    MapPolylineDirective.prototype.GeneratePolylineChangeSet = function (changes) {
        var options = { id: this._id };
        if (changes['Clickable']) {
            options.clickable = this.Clickable;
        }
        if (changes['Draggable']) {
            options.draggable = this.Draggable;
        }
        if (changes['Editable']) {
            options.editable = this.Editable;
        }
        if (changes['Geodesic']) {
            options.geodesic = this.Geodesic;
        }
        if (changes['ShowTooltip']) {
            options.showTooltip = this.ShowTooltip;
        }
        if (changes['StrokeColor']) {
            options.strokeColor = this.StrokeColor;
        }
        if (changes['StrokeOpacity']) {
            options.strokeOpacity = this.StrokeOpacity;
        }
        if (changes['StrokeWeight']) {
            options.strokeWeight = this.StrokeWeight;
        }
        if (changes['Title']) {
            options.title = this.Title;
        }
        if (changes['Visible']) {
            options.visible = this.Visible;
        }
        if (changes['zIndex']) {
            options.zIndex = this.zIndex;
        }
        return options;
    };
    return MapPolylineDirective;
}());
MapPolylineDirective.decorators = [
    { type: core_1.Directive, args: [{
                selector: 'x-map-polyline'
            },] },
];
/** @nocollapse */
MapPolylineDirective.ctorParameters = function () { return [
    { type: polyline_service_1.PolylineService, },
    { type: core_1.ViewContainerRef, },
]; };
MapPolylineDirective.propDecorators = {
    '_infoBox': [{ type: core_1.ContentChild, args: [infobox_1.InfoBoxComponent,] },],
    'Clickable': [{ type: core_1.Input },],
    'Draggable': [{ type: core_1.Input },],
    'Editable': [{ type: core_1.Input },],
    'Geodesic': [{ type: core_1.Input },],
    'Path': [{ type: core_1.Input },],
    'ShowTooltip': [{ type: core_1.Input },],
    'StrokeColor': [{ type: core_1.Input },],
    'StrokeOpacity': [{ type: core_1.Input },],
    'StrokeWeight': [{ type: core_1.Input },],
    'Title': [{ type: core_1.Input },],
    'Visible': [{ type: core_1.Input },],
    'zIndex': [{ type: core_1.Input },],
    'Click': [{ type: core_1.Output },],
    'DblClick': [{ type: core_1.Output },],
    'Drag': [{ type: core_1.Output },],
    'DragEnd': [{ type: core_1.Output },],
    'DragStart': [{ type: core_1.Output },],
    'MouseDown': [{ type: core_1.Output },],
    'MouseMove': [{ type: core_1.Output },],
    'MouseOut': [{ type: core_1.Output },],
    'MouseOver': [{ type: core_1.Output },],
    'MouseUp': [{ type: core_1.Output },],
    'RightClick': [{ type: core_1.Output },],
};
exports.MapPolylineDirective = MapPolylineDirective;
