// angular-maps/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var common_1 = require("@angular/common");
///
/// import module models
///
var info_window_1 = require("./src/models/info-window");
exports.InfoWindow = info_window_1.InfoWindow;
var marker_1 = require("./src/models/marker");
exports.Marker = marker_1.Marker;
var marker_type_id_1 = require("./src/models/marker-type-id");
exports.MarkerTypeId = marker_type_id_1.MarkerTypeId;
var map_type_id_1 = require("./src/models/map-type-id");
exports.MapTypeId = map_type_id_1.MapTypeId;
var layer_1 = require("./src/models/layer");
exports.Layer = layer_1.Layer;
var polygon_1 = require("./src/models/polygon");
exports.Polygon = polygon_1.Polygon;
var polyline_1 = require("./src/models/polyline");
exports.Polyline = polyline_1.Polyline;
var spider_cluster_marker_1 = require("./src/models/spider-cluster-marker");
exports.SpiderClusterMarker = spider_cluster_marker_1.SpiderClusterMarker;
var cluster_placement_mode_1 = require("./src/models/cluster-placement-mode");
exports.ClusterPlacementMode = cluster_placement_mode_1.ClusterPlacementMode;
var cluster_click_action_1 = require("./src/models/cluster-click-action");
exports.ClusterClickAction = cluster_click_action_1.ClusterClickAction;
var bing_layer_1 = require("./src/models/bing/bing-layer");
exports.BingLayer = bing_layer_1.BingLayer;
var bing_spider_cluster_marker_1 = require("./src/models/bing/bing-spider-cluster-marker");
exports.BingSpiderClusterMarker = bing_spider_cluster_marker_1.BingSpiderClusterMarker;
var bing_info_window_1 = require("./src/models/bing/bing-info-window");
exports.BingInfoWindow = bing_info_window_1.BingInfoWindow;
var bing_marker_1 = require("./src/models/bing/bing-marker");
exports.BingMarker = bing_marker_1.BingMarker;
var bing_polygon_1 = require("./src/models/bing/bing-polygon");
exports.BingPolygon = bing_polygon_1.BingPolygon;
var bing_polyline_1 = require("./src/models/bing/bing-polyline");
exports.BingPolyline = bing_polyline_1.BingPolyline;
var google_info_window_1 = require("./src/models/google/google-info-window");
exports.GoogleInfoWindow = google_info_window_1.GoogleInfoWindow;
var google_marker_1 = require("./src/models/google/google-marker");
exports.GoogleMarker = google_marker_1.GoogleMarker;
var google_polygon_1 = require("./src/models/google/google-polygon");
exports.GooglePolygon = google_polygon_1.GooglePolygon;
var google_polyline_1 = require("./src/models/google/google-polyline");
exports.GooglePolyline = google_polyline_1.GooglePolyline;
///
/// import module components
///
var map_1 = require("./src/components/map");
exports.MapComponent = map_1.MapComponent;
var map_marker_1 = require("./src/components/map-marker");
exports.MapMarkerDirective = map_marker_1.MapMarkerDirective;
var infobox_1 = require("./src/components/infobox");
exports.InfoBoxComponent = infobox_1.InfoBoxComponent;
var infobox_action_1 = require("./src/components/infobox-action");
exports.InfoBoxActionDirective = infobox_action_1.InfoBoxActionDirective;
var map_layer_1 = require("./src/components/map-layer");
exports.MapLayerDirective = map_layer_1.MapLayerDirective;
var cluster_layer_1 = require("./src/components/cluster-layer");
exports.ClusterLayerDirective = cluster_layer_1.ClusterLayerDirective;
var map_polygon_1 = require("./src/components/map-polygon");
exports.MapPolygonDirective = map_polygon_1.MapPolygonDirective;
var map_polyline_1 = require("./src/components/map-polyline");
exports.MapPolylineDirective = map_polyline_1.MapPolylineDirective;
///
/// import module services
///
var mapservicefactory_1 = require("./src/services/mapservicefactory");
exports.MapServiceFactory = mapservicefactory_1.MapServiceFactory;
var map_service_1 = require("./src/services/map.service");
exports.MapService = map_service_1.MapService;
var mapapiloader_1 = require("./src/services/mapapiloader");
exports.MapAPILoader = mapapiloader_1.MapAPILoader;
exports.WindowRef = mapapiloader_1.WindowRef;
exports.DocumentRef = mapapiloader_1.DocumentRef;
var infobox_service_1 = require("./src/services/infobox.service");
exports.InfoBoxService = infobox_service_1.InfoBoxService;
var layer_service_1 = require("./src/services/layer.service");
exports.LayerService = layer_service_1.LayerService;
var marker_service_1 = require("./src/services/marker.service");
exports.MarkerService = marker_service_1.MarkerService;
var cluster_service_1 = require("./src/services/cluster.service");
exports.ClusterService = cluster_service_1.ClusterService;
var polygon_service_1 = require("./src/services/polygon.service");
exports.PolygonService = polygon_service_1.PolygonService;
var polyline_service_1 = require("./src/services/polyline.service");
exports.PolylineService = polyline_service_1.PolylineService;
var bing_map_service_factory_1 = require("./src/services/bing/bing-map.service.factory");
exports.BingMapServiceFactory = bing_map_service_factory_1.BingMapServiceFactory;
var bing_map_service_1 = require("./src/services/bing/bing-map.service");
exports.BingMapService = bing_map_service_1.BingMapService;
var bing_map_api_loader_service_1 = require("./src/services/bing/bing-map.api-loader.service");
exports.BingMapAPILoader = bing_map_api_loader_service_1.BingMapAPILoader;
exports.BingMapAPILoaderConfig = bing_map_api_loader_service_1.BingMapAPILoaderConfig;
var bing_infobox_service_1 = require("./src/services/bing/bing-infobox.service");
exports.BingInfoBoxService = bing_infobox_service_1.BingInfoBoxService;
var bing_marker_service_1 = require("./src/services/bing/bing-marker.service");
exports.BingMarkerService = bing_marker_service_1.BingMarkerService;
var bing_layer_service_1 = require("./src/services/bing/bing-layer.service");
exports.BingLayerService = bing_layer_service_1.BingLayerService;
var bing_cluster_service_1 = require("./src/services/bing/bing-cluster.service");
exports.BingClusterService = bing_cluster_service_1.BingClusterService;
var bing_polygon_service_1 = require("./src/services/bing/bing-polygon.service");
exports.BingPolygonService = bing_polygon_service_1.BingPolygonService;
var bing_polyline_service_1 = require("./src/services/bing/bing-polyline.service");
exports.BingPolylineService = bing_polyline_service_1.BingPolylineService;
var google_cluster_service_1 = require("./src/services/google/google-cluster.service");
exports.GoogleClusterService = google_cluster_service_1.GoogleClusterService;
var google_infobox_service_1 = require("./src/services/google/google-infobox.service");
exports.GoogleInfoBoxService = google_infobox_service_1.GoogleInfoBoxService;
var google_layer_service_1 = require("./src/services/google/google-layer.service");
exports.GoogleLayerService = google_layer_service_1.GoogleLayerService;
var google_map_api_loader_service_1 = require("./src/services/google/google-map-api-loader.service");
exports.GoogleMapAPILoader = google_map_api_loader_service_1.GoogleMapAPILoader;
exports.GoogleMapAPILoaderConfig = google_map_api_loader_service_1.GoogleMapAPILoaderConfig;
var google_map_service_factory_1 = require("./src/services/google/google-map.service.factory");
exports.GoogleMapServiceFactory = google_map_service_factory_1.GoogleMapServiceFactory;
var google_map_service_1 = require("./src/services/google/google-map.service");
exports.GoogleMapService = google_map_service_1.GoogleMapService;
var google_marker_service_1 = require("./src/services/google/google-marker.service");
exports.GoogleMarkerService = google_marker_service_1.GoogleMarkerService;
var google_polygon_service_1 = require("./src/services/google/google-polygon.service");
exports.GooglePolygonService = google_polygon_service_1.GooglePolygonService;
var google_polyline_service_1 = require("./src/services/google/google-polyline.service");
exports.GooglePolylineService = google_polyline_service_1.GooglePolylineService;
///
/// define module
///
var MapModule = (function () {
    function MapModule() {
    }
    MapModule.forRoot = function (mapServiceFactory, loader) {
        return {
            ngModule: MapModule,
            providers: [
                mapServiceFactory ? { provide: mapservicefactory_1.MapServiceFactory, useValue: mapServiceFactory } :
                    { provide: mapservicefactory_1.MapServiceFactory, deps: [mapapiloader_1.MapAPILoader, core_1.NgZone], useFactory: bing_map_service_factory_1.BingMapServiceFactoryFactory },
                loader ? { provide: mapapiloader_1.MapAPILoader, useValue: loader } : { provide: mapapiloader_1.MapAPILoader, useFactory: bing_map_service_factory_1.BingMapLoaderFactory },
                mapapiloader_1.DocumentRef,
                mapapiloader_1.WindowRef
            ]
        };
    };
    MapModule.forRootBing = function () {
        return {
            ngModule: MapModule,
            providers: [
                { provide: mapservicefactory_1.MapServiceFactory, deps: [mapapiloader_1.MapAPILoader, core_1.NgZone], useFactory: bing_map_service_factory_1.BingMapServiceFactoryFactory },
                { provide: mapapiloader_1.MapAPILoader, useFactory: bing_map_service_factory_1.BingMapLoaderFactory },
                mapapiloader_1.DocumentRef,
                mapapiloader_1.WindowRef
            ]
        };
    };
    MapModule.forRootGoogle = function () {
        return {
            ngModule: MapModule,
            providers: [
                { provide: mapservicefactory_1.MapServiceFactory, deps: [mapapiloader_1.MapAPILoader, core_1.NgZone], useFactory: google_map_service_factory_1.GoogleMapServiceFactoryFactory },
                { provide: mapapiloader_1.MapAPILoader, useFactory: google_map_service_factory_1.GoogleMapLoaderFactory },
                mapapiloader_1.DocumentRef,
                mapapiloader_1.WindowRef
            ]
        };
    };
    return MapModule;
}());
MapModule.decorators = [
    { type: core_1.NgModule, args: [{
                declarations: [map_layer_1.MapLayerDirective, map_1.MapComponent, map_marker_1.MapMarkerDirective, infobox_1.InfoBoxComponent,
                    infobox_action_1.InfoBoxActionDirective, map_polygon_1.MapPolygonDirective, map_polyline_1.MapPolylineDirective, cluster_layer_1.ClusterLayerDirective],
                imports: [common_1.CommonModule],
                exports: [common_1.CommonModule, map_1.MapComponent, map_marker_1.MapMarkerDirective, map_polygon_1.MapPolygonDirective, map_polyline_1.MapPolylineDirective,
                    infobox_1.InfoBoxComponent, infobox_action_1.InfoBoxActionDirective, map_layer_1.MapLayerDirective, cluster_layer_1.ClusterLayerDirective]
            },] },
];
/** @nocollapse */
MapModule.ctorParameters = function () { return []; };
exports.MapModule = MapModule;
