// angular-maps/dist/src/models/map-type-id.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var MapTypeId;
(function (MapTypeId) {
    /** The aerial map type which uses top-down satellite & airplane imagery. */
    MapTypeId[MapTypeId["aerial"] = 0] = "aerial";
    /** A darker version of the road maps. */
    MapTypeId[MapTypeId["canvasDark"] = 1] = "canvasDark";
    /** A lighter version of the road maps which also has some of the details such as hill shading disabled. */
    MapTypeId[MapTypeId["canvasLight"] = 2] = "canvasLight";
    /** A grayscale version of the road maps. */
    MapTypeId[MapTypeId["grayscale"] = 3] = "grayscale";
    /** The aerial map type including lables */
    MapTypeId[MapTypeId["hybrid"] = 4] = "hybrid";
    /** Displays a blank canvas that uses the mercator map project. It basically removed the base maps layer. */
    MapTypeId[MapTypeId["mercator"] = 5] = "mercator";
    /** Ordnance survey map type (en-gb only). */
    MapTypeId[MapTypeId["ordnanceSurvey"] = 6] = "ordnanceSurvey";
    /** Road map type. */
    MapTypeId[MapTypeId["road"] = 7] = "road";
    /** Provides streetside panoramas from the street level. */
    MapTypeId[MapTypeId["streetside"] = 8] = "streetside";
})(MapTypeId = exports.MapTypeId || (exports.MapTypeId = {}));
